### PR TITLE
New post: Refactoring Rust, Primitive Obsession

### PR DESCRIPTION
**Pull request by IanWhitney**

### CODE REVIEW

File: _posts/2016-08-31-refactoring-rust-primitive-obsession.markdown [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+[`chars` returns an iterator](https://doc.rust-lang.org/std/primitive.str.html#method.chars) and -- just like the `From` trait and type conversion -- Rust has a standard way of adding iterators to your Structs, [`IntoIterator`](https://doc.rust-lang.org/core/iter/trait.IntoIterator.html)
+
+```rust
+impl IntoIterator for Brackets {
+  type Item = char;
+  type IntoIter = ::std::vec::IntoIter<char>;
+
+  fn into_iter(self) -> Self::IntoIter {
+    self.input.chars().collect::<Vec<char>>().into_iter()
```
Comment (Ms2ger) [inline _posts/2016-08-31-refactoring-rust-primitive-obsession.markdown:136]: I would suggest instead implementing `IntoIterator` like this:

``` rust
impl<'a> IntoIterator for &'a Brackets {
    type Item = char;
    type IntoIter = ::std::str::Chars<'a>;

    fn into_iter(self) -> Self::IntoIter {
        self.input.chars()
    }
}
```

This avoids allocating a vector for no particular reason.

Comment (IanWhitney) [reply _posts/2016-08-31-refactoring-rust-primitive-obsession.markdown:136]: Thanks! I tried that but could not get the lifetime annotations correct and decided it in favor of the simpler-looking (if less efficient) vector approach.

Comment (IanWhitney) [reply _posts/2016-08-31-refactoring-rust-primitive-obsession.markdown:136]: I've updated the post to link to your non-vector impl.

**Comment by hroi** [comment]
More generic:

``` rust
struct Brackets<S> {
    input: S,
}

impl<S> From<S> for Brackets<S> {
    fn from(input: S) -> Self {
        Brackets { input: input }
    }
}

impl<S> IntoIterator for Brackets<S> where S: AsRef<str> {
    type Item = char;
    type IntoIter = ::std::vec::IntoIter<char>;

    fn into_iter(self) -> Self::IntoIter {
        self.input.as_ref().chars().collect::<Vec<char>>().into_iter()
    }
}

pub fn are_balanced<S>(input: S) -> bool
    where S: AsRef<str>
{
    let input = input.as_ref();
    let brackets = Brackets::from(input);
    let mut unmatched_brackets: Vec<char> = Vec::new();

    for current_bracket in brackets.into_iter() {
        if let Some(most_recent_unmatched) = unmatched_brackets.pop() {
            let is_matched = match (most_recent_unmatched, current_bracket) {
                ('[', ']') => true,
                ('{', '}') => true,
                ('(', ')') => true,
                _ => false,
            };
            if !is_matched {
                unmatched_brackets.push(most_recent_unmatched);
                unmatched_brackets.push(current_bracket);
            }
        } else {
            unmatched_brackets.push(current_bracket);
        }
    }

    unmatched_brackets.is_empty()
}

fn main() {
    assert!(are_balanced("{[]}".to_string()));
    assert!(!are_balanced("[{]}")); // works on primitive strs as well
}
```

**Comment by IanWhitney** [comment]
@hroi, thanks. I'm guessing generics will show up later in this series of posts, a little early to introduce them in the first entry.

**Comment by wigy-opensource-developer** [comment]
I love the idea of this series of articles. Keep on the good work! Do you have a schedule for the next article?

**Comment by IanWhitney** [comment]
@wigy-opensource-developer I'm glad you like it. I wish I could commit to a schedule right now, but life is making that hard! As soon as I'm able.

**Pull request merged**